File: sales/sales.js
frappe.ui.form.on('Sales', {
    refresh(frm) {
        if (frm.doc.docstatus === 1) {
            frm.add_custom_button('Update', () => {
                let table_data = frm.doc.choose_items.map(row => ({
                    existing_row_name: row.name,
                   
                    name1:  row.name1,  
                    qty: row.qty,
                    rate: row.rate,
                    amount: row.amount
                }));

                let dialog = new frappe.ui.Dialog({
                    title: 'Update',
                    fields: [
                        {
                            label: 'Choose Items',
                            fieldname: 'choose_items',
                            fieldtype: 'Table',
                            cannot_add_rows: false,
                            in_place_edit: true,
                            data: table_data,
                            fields: [
                                {
                                    label: 'Item',
                                    fieldname: 'name1',
                                    fieldtype: 'Link',
                                    options: 'Items',
                                    in_list_view: true
                                },
                                {
                                    label: 'Rate',
                                    fieldname: 'rate',
                                    fieldtype: 'Int',
                                    in_list_view: true
                                },
                                {
                                    label: 'Qty',
                                    fieldname: 'qty',
                                    fieldtype: 'Int',
                                    in_list_view: true
                                },{
                                    fieldname : 'existing_row_name',
                                    fieldtype: 'Data'
                                }

                            ]
                        }
                    ],
                    primary_action_label: 'Update',
                    primary_action(values) {
                        


                        if(!values.choose_items) {
                            frappe.msgprint('Please add at least one item to update.');
                            return;
                        }
                        for (let item of values.choose_items) {
                            if (!item.name1 || !item.qty || !item.rate || item.qty <= 0 || item.rate <= 0) {
                                frappe.msgprint('Please fill all fields for each item.');
                                return;
                            }
                        }





                        
                        dialog.hide();
                        frappe.call({   
                            method: 'salesapp.salesapp.doctype.sales.sales.update',
                            args: {
                                docname: frm.doc.name,
                                updated: JSON.stringify(values.choose_items)
                            },
                            callback: function (r) {
                                if (r.message) {
                                    frappe.show_alert({
                                        message: r.message,
                                        indicator: 'green'
                                    });
                                    frm.reload_doc();
                                } else {
                                    frappe.show_alert({
                                        message: 'Error updating sales',
                                        indicator: 'red'
                                    });
                                }
                            }
                        });
                        
                    }
                });

                dialog.show();
            });
        }
    },
    name1: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (row.name1) {
            frappe.db.get_value('Items', row.name1, 'rate', (r) => {
                if (r && r.rate !== undefined) {
                    row.rate = r.rate;
                    frappe.model.set_value(cdt, cdn, 'rate', r.rate);
                    row.amount = (r.rate || 0) * (row.qty || 0);
                    frm.refresh_field('choose_items');
                    update_total_amount(frm);
                }
            });
        }
    },

    qty: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        row.amount = (row.rate || 0) * (row.qty || 0);
        frm.refresh_field('choose_items');
        update_total_amount(frm);
    },
    
    rate: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        row.amount = (row.rate || 0) * (row.qty || 0);
        frm.refresh_field('choose_items');
        update_total_amount(frm);
    },

    
});

function update_total_amount(frm) {
    let total = 0;
    (frm.doc.choose_items || []).forEach(row => {
        total += row.amount || 0;
    });
    frm.set_value('total_amount', total);
}

File: sales/sales.py
import frappe
from frappe.model.document import Document
import json 


class Sales(Document):
    def validate(self):
        if not self.choose_items:
            frappe.throw("Sales transaction should not be empty")

        self.total_amount = 0
        for i in self.choose_items:
            

            i.amount = i.qty * i.rate
            self.total_amount += i.amount

    def on_update_after_submit(self):
        self.total_amount = 0
        for i in self.choose_items:
            i.amount = i.qty * i.rate
            self.total_amount += i.amount


@frappe.whitelist()
def update(docname, updated):
    import json
    updated = json.loads(updated)
    existing_row =frappe.get_all("Sales Items", filters={"parent": docname}, fields=["name"])
    

    total_amount = 0

    updated_rows = set()

    for item in updated:
        name1 = item.get("name1")
        qty = float(item.get("qty"))
        rate = float(item.get("rate"))
        amount = qty * rate
        total_amount += amount
        row_name = item.get("name")

       

        if row_name and row_name in existing_row:
        
            
            frappe.db.set_value("Sales Items", row_name, {
                "name1": name1,
                "qty": qty,
                "rate": rate,
                "amount": amount
            })
            updated_rows.add(row_name)
        else:
            
            new_row = frappe.new_doc("Sales Items")
            new_row.parent = docname
            new_row.parenttype = "Sales"
            new_row.parentfield = "choose_items"
            new_row.name1 = name1
            new_row.qty = qty
            new_row.rate = rate
            new_row.amount = amount
            
            new_row.insert()
            updated_rows.add(new_row.name)

    for key in existing_row:
        if key.name not in updated_rows:
            frappe.delete_doc("Sales Items", key.name)

    

    
    # # Delete any orphan items (optional safety cleanup)
    # frappe.db.sql("""
    #     DELETE FROM `tabSales Items`
    #     WHERE parent IS NULL OR parent = ''
    # """)

    # Update total amount in Sales
    frappe.db.set_value("Sales", docname, "total_amount", total_amount)

    frappe.db.commit()

    return "Sales updated successfully"

